### ISSUE TITLE
Fix Request: BufferFilesNewStream rule support for BufferedReader and BufferedWriter, and false positives with methods using buffers

### ISSUE THREAD

**Issue by dylan-tao** [jborgers/PMD-jPinpoint-rules#539]
BufferFilesNewStream rule support for BufferedReader and BufferedWriter, missing character stream for current check condition.

```
public static BufferedReader newBufferedReader(File file, Charset charset) throws IOException {
    Assertions.notExists(file, "file is required");
    return new BufferedReader(new InputStreamReader(Files.newInputStream(file.toPath()), charset), BUFFER_SIZE);
  }

  public static BufferedWriter newBufferedWriter(String pathname, Charset charset) throws IOException {
    Assertions.isBlank(pathname, "pathname is required");
    return new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(Paths.of(pathname)), charset), BUFFER_SIZE);
  }
```

**Comment by dylan-tao** [comment]
Fixed rule:
```
<![CDATA[
//PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newOutputStream')]
[ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedOutputStream')]) = 0
and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedWriter')]) = 0
]]
,
//PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newInputStream')]
[ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedInputStream')]) = 0
and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedReader')]) = 0
and (not(.//PrimaryPrefix/Name[pmd-java:typeIs('org.apache.commons.io.IOUtils') and starts-with(@Image, 'IOUtils.read')]))
]]
 ]]>
```

**Comment by dylan-tao** [comment]
> Fixed rule:
> 
> ```
> <![CDATA[
> //PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newOutputStream')]
> [ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedOutputStream')]) = 0
> and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedWriter')]) = 0
> ]]
> ,
> //PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newInputStream')]
> [ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedInputStream')]) = 0
> and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedReader')]) = 0
> and (not(.//PrimaryPrefix/Name[pmd-java:typeIs('org.apache.commons.io.IOUtils') and starts-with(@Image, 'IOUtils.read')]))
> ]]
>  ]]>
> ```

Rule new issues： InputStream.readAllBytes() and InputStream.transferTo(OutputStream) included buffer，need ignore

**Comment by dylan-tao** [comment]
> > Fixed rule:
> > ```
> > <![CDATA[
> > //PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newOutputStream')]
> > [ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedOutputStream')]) = 0
> > and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedWriter')]) = 0
> > ]]
> > ,
> > //PrimaryExpression/PrimaryPrefix/Name[pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newInputStream')]
> > [ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedInputStream')]) = 0
> > and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedReader')]) = 0
> > and (not(.//PrimaryPrefix/Name[pmd-java:typeIs('org.apache.commons.io.IOUtils') and starts-with(@Image, 'IOUtils.read')]))
> > ]]
> >  ]]>
> > ```
> 
> Rule new issues： InputStream.readAllBytes() and InputStream.transferTo(OutputStream) included buffer，need ignore

Final fixed rule：
```
<![CDATA[
//PrimaryExpression/PrimaryPrefix/Name[(pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newOutputStream'))
and not(ancestor::MethodDeclaration[.//PrimaryExpression/PrimaryPrefix/Name[@Image = 'transferTo']])]
[ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedOutputStream')]) = 0
and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedWriter')]) = 0
]]
,
//PrimaryExpression/PrimaryPrefix/Name[(pmd-java:typeIs('java.nio.file.Files') and ends-with(@Image,'.newInputStream'))
and not(ancestor::MethodDeclaration[.//PrimaryExpression/PrimaryPrefix/Name[@Image = 'readAllBytes']])
and not(ancestor::MethodDeclaration[.//PrimaryExpression/PrimaryPrefix/Name[@Image = 'transferTo']])]
[ancestor::MethodDeclaration[count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedInputStream')]) = 0
and count(.//AllocationExpression[pmd-java:typeIs('java.io.BufferedReader')]) = 0
and (not(.//PrimaryPrefix/Name[pmd-java:typeIs('org.apache.commons.io.IOUtils') and starts-with(@Image, 'IOUtils.read')]))
]]
]]>
```

**Comment by jborgers** [comment]
Thanks for reporting. Please suggest improvements on the pmd7 branch, since we are not improving the pmd6 (currently still master) branch anymore.

**Comment by jborgers** [comment]
Test cases:
```
class ReaderTestCases {
    public static BufferedReader newBufferedReader(Path path) throws IOException {
        return new BufferedReader(new InputStreamReader(Files.newInputStream(path))); // <- false positive
    }

    public static BufferedWriter newBufferedWriter(Path path) throws IOException {
        return new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(path))); // <- false positive
    }
}

class BufferedTestCases {
    void goodTransfer(InputStream in, Path pathOut) throws IOException {
        in.transferTo(Files.newOutputStream(pathOut)); // <- false positive
    }
    void goodTransfer2(Path pathIn, OutputStream out) throws IOException {
        Files.newInputStream(pathIn).transferTo(out); // <- false positive
    }
    byte[] goodReadAll(String pathIn) throws IOException {
        InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
        return in.readAllBytes();
    }
    byte[] goodReadN(String pathIn) throws IOException {
        InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
        return in.readNBytes(4096);
    }
   int goodReadArray(String pathIn) throws IOException {
        InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
        byte[] buf = new byte[4096];
        return in.read(buf);
    }
   int goodReadArray2(String pathIn) throws IOException {
        InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
        byte[] buf = new byte[4096];
        return in.read(buf, 0, 4096);
    }
    void goodCopy(Path pathIn, Path pathOut) throws IOException {
        Files.copy(pathIn, Files.newOutputStream(pathOut)); // <- false positive
    }
}
```

**Comment by dylan-tao** [comment]
> Test cases:
> 
> ```
> class ReaderTestCases {
>     public static BufferedReader newBufferedReader(Path path) throws IOException {
>         return new BufferedReader(new InputStreamReader(Files.newInputStream(path))); // <- false positive
>     }
> 
>     public static BufferedWriter newBufferedWriter(Path path) throws IOException {
>         return new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(path))); // <- false positive
>     }
> }
> 
> class BufferedTestCases {
>     void goodTransfer(InputStream in, Path pathOut) throws IOException {
>         in.transferTo(Files.newOutputStream(pathOut)); // <- false positive
>     }
>     void goodTransfer2(Path pathIn, OutputStream out) throws IOException {
>         Files.newInputStream(pathIn).transferTo(out); // <- false positive
>     }
>     byte[] goodReadAll(String pathIn) throws IOException {
>         InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
>         return in.readAllBytes();
>     }
>     byte[] goodReadN(String pathIn) throws IOException {
>         InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
>         return in.readNBytes(4096);
>     }
>    int goodReadArray(String pathIn) throws IOException {
>         InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
>         byte[] buf = new byte[4096];
>         return in.read(buf);
>     }
>    int goodReadArray2(String pathIn) throws IOException {
>         InputStream in = Files.newInputStream(Paths.get(pathIn)); // <- false positive
>         byte[] buf = new byte[4096];
>         return in.read(buf, 0, 4096);
>     }
>     void goodCopy(Path pathIn, Path pathOut) throws IOException {
>         Files.copy(pathIn, Files.newOutputStream(pathOut)); // <- false positive
>     }
> }
> ```

Okay, I will stay at pmd6 for about 6 months, and there are still a lot of custom check rules that need to be upgraded (but this is not the main work at the moment). After the internal upgrade is completed, the jPinpoint version will be upgraded. Thanks！
